mne/preprocessing/_regress.py

```diff
@@ -53,35 +74,244 @@ def regress_artifact(inst, picks=None, picks_artifact='eog', betas=None,
     ----------
     .. footbibliography::
     """  # noqa: E501
-    _check_preload(inst, 'regress')
-    _validate_type(inst, (BaseEpochs, BaseRaw), 'inst', 'Epochs or Raw')
-    picks = _picks_to_idx(inst.info, picks, none='data')
-    picks_artifact = _picks_to_idx(inst.info, picks_artifact)
-    if np.in1d(picks_artifact, picks).any():
-        raise ValueError('picks_artifact cannot be contained in picks')
-    inst = inst.copy() if copy else inst
-    artifact_data = inst._data[..., picks_artifact, :]
-    ref_data = artifact_data - np.mean(artifact_data, -1, keepdims=True)
-    if ref_data.ndim == 3:
-        ref_data = ref_data.transpose(1, 0, 2).reshape(len(picks_artifact), -1)
-    cov = np.dot(ref_data, ref_data.T)
-    # process each one separately to reduce memory load
-    betas_shape = (len(picks), len(picks_artifact))
     if betas is None:
-        betas = np.empty(betas_shape)
-        estimate = True
+        model = EOGRegression(picks=picks, exclude=exclude,
+                              picks_artifact=picks_artifact, taa=taa,
+                              proj=proj)
+        model.fit(inst)
     else:
-        estimate = False
-        betas = np.asarray(betas, dtype=float)
-        _check_option('betas.shape', betas.shape, (betas_shape,))
-    for pi, pick in enumerate(picks):
-        this_data = inst._data[..., pick, :]  # view
-        orig_shape = this_data.shape
-        if estimate:
-            # subtract mean over time from every trial/channel
+        # Create an EOGRegression object and load the given betas into it.
+        picks = _picks_to_idx(inst.info, picks, exclude=exclude, none='data')
+        picks_artifact = _picks_to_idx(inst.info, picks_artifact)
+        want_betas_shape = (len(picks), len(picks_artifact))
+        _check_option('betas.shape', betas.shape, (want_betas_shape,))
+        model = EOGRegression(picks, picks_artifact, taa=taa, proj=proj)
+        model.info = pick_info(inst.info, picks)
+        model.coef_ = betas
+        model._picks = model.info['ch_names']
+        model._exclude = exclude
+        model._picks = model.info['ch_names']
+        model._picks_artifact = [inst.ch_names[ch] for ch in picks_artifact]
```

Comment:
having to hack private attributes outside the class definition is a code smell for me. Why is picks_artifact not the same as what is passed in the constructor? Can we try to do it by only passing constructor parameters and do some validation in there?

Reply:
This is the second time you have asked me about this, which indicates that this isn't obvious. So I bit the bullet and refactored so that `.picks` and `.picks_artifact` are just copies of the supplied parameters. The reason for storing picks as one channel name per regression coefficient was that we could re-match regression coefficients to channels when we `.apply()`. After the refactor we can still deal with extra channels being present (the most likely scenario, e.g. a stim channel being present), but we have lost the ability to compensate when the ordering of the channels (an unlikely scenario).